Eventbrite block: add test coverage and add input trim

ramonjd opened the pull request:
Fixes https://github.com/Automattic/jetpack/issues/19088

#### Changes proposed in this Pull Request:
This PR:

- adds basic test coverage / fixtures for block and deprecations
- fixes a small input sanitization bug reported by https://github.com/Automattic/jetpack/issues/19088 and adds basic coverage for this test by abstracting trim method.

#### Jetpack product discussion
p1HpG7-b3G-p2


#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:

See https://github.com/Automattic/jetpack/issues/19088  for testing the input bug

Run the unit tests:
```
cd projects/plugins/jetpack
yarn fixtures:test extensions/blocks/eventbrite/test
```

ramonjd committed:
[not verified] initial commit: adding fixtures and addressing bug in https://github.com/Automattic/jetpack/issues/19088

ramonjd committed:
[not verified] Adding deprecated fixtures

ramonjd committed:
[not verified] Revert space change
add changelog

ramonjd commented:
It seems as though the fixture [jetpack__eventbrite__deprecated-2.html ](https://github.com/Automattic/jetpack/blob/7e2f3ab16b3c7345bb42e34194f93d8661545624/projects/plugins/jetpack/extensions/blocks/eventbrite/test/fixtures/jetpack__eventbrite__deprecated-2.html) throws the following error in the v2's use of [createBlock](https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/extensions/blocks/eventbrite/deprecated/v2/index.js#L70)

```
  ● Test block content parsing › all content versions parse correctly for block jetpack/eventbrite

    TypeError: Cannot read property 'attributes' of undefined

      69 | 
      70 |              const newInnerBlocks = [
    > 71 |                      createBlock( 'jetpack/button', {
         |                      ^

```

Not quite sure why yet, but it fails fairly consistently. Logging in the built version of [createBlock](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/factory.js#L47) informs me that `getBlockType('name')` returns `undefined`. 

Maybe the block isn't registered in the tests?

Updating the `@wordpress/blocks` package removes the error, but breaks every fixture test in the repo 🤣 

cc @aaronrobertshaw @glendaviesnz to ask for their advice 🙇

aaronrobertshaw commented:
@ramonjd At first glance I believe the issue is that you need to add the button block type definition for the inner block to the blocks array in `validate.js`.

ramonjd committed:
[not verified] Importing button block type definitions into the tests.

ramonjd committed:
Refreshing fixtures after importing button settings

ramonjd committed:
[not verified] Exporting controls
Adding basic tests for controls, utils and edit
Covering the bug fix by extracting a standard trim/normalize method

ramonjd committed:
Resolving comment doc lint issues (must have hyphen before @param desc)

ramonjd committed:
[not verified] 'url' is assigned a value but never used  no-unused-vars

andrewserong commented on projects/plugins/jetpack/extensions/blocks/eventbrite/test/fixtures/jetpack__eventbrite-modal__deprecated-1.serialized.html:
```diff
@@ -0,0 +1,3 @@
+<!-- wp:jetpack/eventbrite {"url":"https://www.eventbrite.com.au/e/food-hub-unsw-tickets-144144099883","eventId":144144099883,"style":"modal"} -->
+<div class="wp-block-jetpack-eventbrite"></div>
```
It looks like the v1 deprecation is missing a migration to migrate the button to a `jetpack-button` block. This serialized output is missing the following from within the `div`:

```
<!-- wp:jetpack/button {"element":"a","uniqueId":"eventbrite-widget-id","text":"Register"} /-->
```

ramonjd replied:
Thanks @andrewserong! I'll take a look this afternoon. 

I'll grok deprecations before the end of the century, I'm sure. 😄

ramonjd replied:
So in aa060ec2bd1e5965c5532dd68c8d6636723c247f I copied the migrate function from v2 deprecation to v1. The only change was the omission of the deprecated attribute `useModal`.

It made me think: v1 and v2 deprecations are identical, save for the `useModal` attribute. 

Is it worth trying to combine them into one deprecation, and migrating the `userModal` attribute if found? 

I don't know the answer, but it seems like a huge duplication.

andrewserong replied:
Great point. I agree, I think it’s worth seeing if they can be combined to reduce the duplication, given that they’re so similar.

andrewserong replied:
Taking a look at these deprecations again, I think I don't quite understand what the difference is between the original markup that would've been generated by v1 vs v2. At the moment the source HTML for each of these deprecation fixtures (using the modal) looks to be the same.

If the markup expected by each of the deprecations is identical, then we likely don't need two separate deprecations. If it is different, then we might need to update the two source deprecation fixtures so that each of those deprecations is targeted differently? (The files: `jetpack__eventbrite-modal__deprecated-1.html` and `jetpack__eventbrite-modal__deprecated-2.html`) 🤔

ramonjd replied:
> Taking a look at these deprecations again, I think I don't quite understand what the difference is between the original markup that would've been generated by v1 vs v2.

Yeah, it's a bit confusing. I can't really see, aside from `attributes.useModal` where they differ now. 

The last deprecation was introduced in https://github.com/Automattic/jetpack/pull/15918 I think. There might be some clues there.

aaronrobertshaw replied:
After reading over the deprecations, I think I'm starting to get a clearer picture of what's needed. That said, I haven't tried tweaking the deprecations yet so take the following with a grain of salt.

My understanding is:
- V1 and v2 deprecations would match the same HTML markup whether that is for the modal or inline displays.
- The difference between v1 and v2 was the switch from `useModal` to `style`, no markup change.
- Because there was no change in markup between v1 and v2 we should be able to merge the deprecations
    - the `migrate` function will just need to handle the two cases; with `useModal` or not.
    - when v2 was added v1 should have migrated to the button block so they both produced the same result.
- Due to the fork in rendering we could still have non-modal style markup matching so need the `isEligible` check.
- It also looks to me like a deprecation may have been missed when the non-modal markup was changed in https://github.com/Automattic/jetpack/pull/14558.
    - Given the low insertion rate of the block and the small number of months between the above change and v2 I think we can ignore that.
- Even if we merge the deprecations, we'll want the fixtures to still cover the same possibilities.
    - Hopefully with some clear naming and comments to help our future selves 🙂 
- Reducing the number of deprecations does give a tiny performance bump as well.

Feel free to point out whatever I might have missed.

To move forward my suggestion would be to:
1. Merge v1 & v2
2. Update the fixtures as needed to cover each use-case the deprecations handle
3. Document what's going on inline as much as possible

ramonjd replied:
Thanks for taking the time to review and confirm all this @aaronrobertshaw. 🙇 

I've made an attempt at this in 7b9d77a3649df8acf0721ddd7ad804de4ecae790

If folks think it's all too much noise for this PR I can extract it and do a follow up.

andrewserong reviewed (changes requested):
Nice work @ramonjd, the input trim bug fix is working nicely for me, and thanks for including all the different variations of the fixtures tests. I confirmed that they're all passing, but I think the v1 deprecation needs to be fixed up to handle migrating the inner button to the `jetpack/button` block. At the moment it appears to strip the inner content out, I've left a comment on the v1 modal serialized html.

ramonjd committed:
[not verified] Modifying v1 deprecation so that we're generating the expected jetpack/button block in the save function.

aaronrobertshaw commented on projects/plugins/jetpack/extensions/blocks/eventbrite/test/fixtures/jetpack__eventbrite__deprecated-1.parsed.json:
```diff
@@ -0,0 +1,23 @@
+            "\n<a class=\"wp-block-jetpack-eventbrite eventbrite__direct-link\" href=\"https://www.eventbrite.com.au/e/the-fever-tree-gin-and-tonic-festival-2021-tickets-140737974069\">https://www.eventbrite.com.au/e/the-fever-tree-gin-and-tonic-festival-2021-tickets-140737974069</a>\n"
+        ]
+    },
+    {
+        "blockName": null,
+        "attrs": {},
+        "innerBlocks": [],
+        "innerHTML": "\n",
+        "innerContent": [
+            "\n"
+        ]
+    }
```
This is the result of a blank line at the end of the `.html` fixture file.

ramonjd replied:
😱

ramonjd replied:
Maybe I have a fix here: https://github.com/Automattic/jetpack/pull/19180

ramonjd committed:
[not verified] # This is a combination of 2 commits.

Merging v1 and v2 deprecations since there are no markup changes between the two. The difference is the use of the `useModal` (bool) attribute, which is now deprecated in favour of `style` (string).
Updated the deprecation fixtures to reflect layout variations and deprecated attributes.
Added inline comments.

ramonjd committed:
[not verified] Trimming fixture html before generation removes the hanging block in the parsed JSON

aaronrobertshaw committed:
Regenerate fixtures to remove empty parsed blocks

aaronrobertshaw committed:
Remove placeholder comment

aaronrobertshaw committed:
Fix linting errors

aaronrobertshaw commented:
@ramonjd I've just pushed a couple of small commits. 

The main one was just to regenerate the block fixtures as the tests were failing after the trimming of the seed file. It then no longer matched the existing .json file. The others were just removing unneeded comment and linting errors.

aaronrobertshaw reviewed (approved):
Nicely done 👍 

I think this one should be good to go now. The tests are all passing for me and the deprecations look solid.

andrewserong reviewed (approved):
LGTM to me too, thanks for combining the deprecations, that looks much neater 👍

jeherve reviewed (approved):
This tests well for me. 👍

ramonjd commented:
r222950-wpcom
